### PR TITLE
Model and store column lineage in Marquez DB

### PR DESCRIPTION
**Pull request by mzareba382**
### Problem & solution

This PR adds a column-level lineage representation and API endpoint to retrieve column-level lineage data from Marquez's database. It is based on OpenLineage's [column-level lineage facet](https://github.com/OpenLineage/OpenLineage/blob/main/spec/facets/ColumnLineageDatasetFacet.md)

Column-level lineage data is being stored in separate table with following fields:
```
CREATE TABLE column_lineage (
  output_dataset_version_uuid   uuid REFERENCES dataset_versions(uuid), -- allows join to run_id
  output_dataset_field_uuid     uuid REFERENCES dataset_fields(uuid),
  input_dataset_version_uuid    uuid REFERENCES dataset_versions(uuid), -- speed up graph column lineage graph traversal
  input_dataset_field_uuid      uuid REFERENCES dataset_fields(uuid),
  transformation_description    VARCHAR(255) NOT NULL,
  transformation_type           VARCHAR(255) NOT NULL,
  created_at                    TIMESTAMP NOT NULL,
  updated_at                    TIMESTAMP NOT NULL,
  UNIQUE (output_dataset_version_uuid, output_dataset_field_uuid, input_dataset_version_uuid, input_dataset_field_uuid)
);
```

Relevant tickets are:
- https://github.com/MarquezProject/marquez/issues/2045
- https://github.com/MarquezProject/marquez/pull/2077

Solution: 
- [x] Column-level lineage DB table with its DAO class
- [x] Update relevant methods in OpenLineageDao class to upsert table on OpenLineage metadata arrival.
- [x] Output_column_name and input_field should be linked via uuid's from dataset_fields table. Normalize this table to represent lineage as references in database.
- [x] Implement tests in OpenLineageDaoTests to check if column-level lineage data is properly retrieved from facets and saved to db.

### Checklist

- [x] You've [signed-off](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#sign-your-work) your work
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [x] You've updated the [`CHANGELOG.md`](https://github.com/MarquezProject/marquez/blob/main/CHANGELOG.md#unreleased) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned your `.sql` database schema migration according to [Flyway's naming convention](https://flywaydb.org/documentation/concepts/migrations#naming) (_if relevant_)
- [ ] You've included a [header](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#copyright--license) in any source code files (_if relevant_)

### CODE REVIEW

File: api/src/test/java/marquez/db/OpenLineageDaoTest.java [hunk 1]
```diff
@@ -94,6 +95,41 @@ void testUpdateMarquezModelLifecycleStateChangeFacet() {
...
+    JobFacet jobFacet = new JobFacet(null, null, null, LineageTestUtils.EMPTY_MAP);
+    UpdateLineageRow writeJob =
+        LineageTestUtils.createLineageRow(
+            dao, WRITE_JOB_NAME, "COMPLETE", jobFacet, Arrays.asList(), Arrays.asList(dataset));
+
+    assertThat(
+            writeJob.getOutputs().get().stream().toList().stream()
+                .findAny()
+                .orElseThrow()
+                .getColumnLineageRows())
```
Comment (pawel-big-lebowski) [inline api/src/test/java/marquez/db/OpenLineageDaoTest.java:128]: Great, this is how the test should look like. 
I think it's worth testing column lineage row columns either.

File: api/src/main/java/marquez/db/ColumnLevelLineageDao.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            dataset_version_uuid, output_column_name, input_field)
+        .orElseThrow();
+  }
+
+  @SqlQuery("SELECT * FROM column_level_lineage WHERE dataset_version_uuid = :datasetVersionUuid")
+  Optional<ColumnLevelLineageRow> findColumnLevelLineageByDatasetVersionColumnAndInput(
+      UUID datasetVersionUuid, String outputColumnName, String inputField);
+
+  @SqlUpdate(
+      "INSERT INTO column_level_lineage ("
```
Comment (mobuchowski) [inline api/src/main/java/marquez/db/ColumnLevelLineageDao.java:46]: New SQL should use Java 17's text blocks.

Comment (mzareba382) [reply api/src/main/java/marquez/db/ColumnLevelLineageDao.java]: Done.

File: api/src/test/java/marquez/db/OpenLineageDaoTest.java (hunk 1 above)
Comment (mzareba382) [reply api/src/test/java/marquez/db/OpenLineageDaoTest.java:137]: Done.

File: api/src/main/java/marquez/db/OpenLineageDao.java [hunk 3]
```diff
@@ -649,7 +661,85 @@ default DatasetRecord upsertLineageDataset(
...
+            outputColumn -> {
+              Optional<DatasetFieldRow> outputField =
+                  datasetFields.stream()
+                      .filter(dfr -> dfr.getName().equals(outputColumn.getName()))
+                      .findAny();
+
+              if (outputField.isEmpty()) {
+                Logger log = LoggerFactory.getLogger(OpenLineageDao.class);
+                log.error(
+                    "Cannot produce column lineage for missing output field in output dataset: %s",
```
Comment (pawel-big-lebowski) [inline api/src/main/java/marquez/db/OpenLineageDao.java:706]: {} should be used

**Comment by wslulciuc** [comment]
@mzareba382 I was wondering if this PR is now outdated? /cc @mobuchowski, @pawel-big-lebowski

File: api/src/main/java/marquez/db/ColumnLevelLineageDao.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          updated_at
+          ) VALUES <values>
+          ON CONFLICT (output_dataset_version_uuid, output_dataset_field_uuid, input_dataset_version_uuid, input_dataset_field_uuid)
+          DO UPDATE SET
+          transformation_description = EXCLUDED.transformation_description,
+          transformation_type = EXCLUDED.transformation_type,
+          updated_at = EXCLUDED.updated_at
+          RETURNING *
+          """)
+  void doUpsertColumnLevelLineageRow(
```
Comment (mobuchowski) [inline api/src/main/java/marquez/db/ColumnLevelLineageDao.java:79]: If you're `RETURNING *` the type here shouldn't be void. You can also get rid of 
```
    doUpsertColumnLevelLineageRow(rows.toArray(new ColumnLevelLineageRow[0]));
    return findColumnLevelLineageByDatasetVersionColumnAndOutputDatasetField(
        outputDatasetVersionUuid, outputDatasetFieldUuid);
```
and just return result of `doUpsertColumnLevelLineageRow`.

**Review by mobuchowski — CHANGES_REQUESTED** [review]

File: api/src/main/java/marquez/db/OpenLineageDao.java [hunk 5]
```diff
@@ -649,7 +661,85 @@ default DatasetRecord upsertLineageDataset(
...
+        .flatMap(list -> list.stream())
+        .flatMap(
+            outputColumn -> {
+              Optional<DatasetFieldRow> outputField =
+                  datasetFields.stream()
+                      .filter(dfr -> dfr.getName().equals(outputColumn.getName()))
+                      .findAny();
+
+              if (outputField.isEmpty()) {
+                Logger log = LoggerFactory.getLogger(OpenLineageDao.class);
```
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/OpenLineageDao.java:704]: it's not a class, but an interface.

File: api/src/main/java/marquez/db/OpenLineageDao.java [hunk 6]
```diff
@@ -649,7 +661,85 @@ default DatasetRecord upsertLineageDataset(
...
+  private List<ColumnLevelLineageRow> upsertColumnLineage(
+      UUID runUuid,
+      Dataset ds,
+      Instant now,
+      List<DatasetFieldRow> datasetFields,
+      ColumnLevelLineageDao columnLevelLineageDao,
+      DatasetFieldDao datasetFieldDao,
+      DatasetVersionRow datasetVersionRow) {
+    // get all the fields related to this particular run
+    List<InputFieldData> runFields = datasetFieldDao.findInputFieldsDataAssociatedWithRun(runUuid);
```
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/OpenLineageDao.java:689]: * getting it from DB returns UUIDs'
* some info, like inputs, can be in START event.

File: api/src/main/java/marquez/db/ColumnLevelLineageDao.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                "outputDatasetFieldUuid",
+                "inputDatasetVersionUuid",
+                "inputDatasetFieldUuid",
+                "transformationDescription",
+                "transformationType",
+                "createdAt",
+                "updatedAt"
+              },
+              value = "values")
+          ColumnLevelLineageRow... rows);
```
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/ColumnLevelLineageDao.java:92]: Varargs will not work with list and `new ColumnLevelLineageRow[0]` is necessary to provide a type of returned array. Not sure if this is what you meant.

File: api/src/main/java/marquez/db/ColumnLevelLineageDao.java (hunk 4 above)
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/ColumnLevelLineageDao.java]: I removed `RETURNING *` clause. 
Returning gets only the modified rows while we need to return `DatasetRecord` with all the fields and a whole column lineage. That's why two separate queries are run.

File: api/src/main/java/marquez/db/ColumnLevelLineageDao.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mobuchowski) [reply api/src/main/java/marquez/db/ColumnLevelLineageDao.java]: I mean to change the method signature from varargs to a list.

BindBeanList can deal with Iterables: https://jdbi.org/apidocs/org/jdbi/v3/sqlobject/customizer/BindBeanList.html

File: api/src/main/java/marquez/db/ColumnLevelLineageDao.java (hunk 4 above)
Comment (mobuchowski) [reply api/src/main/java/marquez/db/ColumnLevelLineageDao.java]: Ok, makes sense.

File: api/src/main/java/marquez/db/ColumnLevelLineageDao.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/ColumnLevelLineageDao.java]: Fab. Great to know that.

**Review by mobuchowski — APPROVED** [review]

**Review by wslulciuc — APPROVED** [review]
Great work on modeling the `column_lineage` table and separating the column-level lineage facet into a separate table (similar to the design proposed in #2076). Also great to see extensive testing around this feature! 💯 🥇

**Pull request merged**